task1/task1.js

```diff
@@ -0,0 +1,160 @@
+//a
+Array.prototype.map2 = function (callback, thisArg) {
+
+    if (this === null) {
+      throw new Error("Cant iterate over undefined or null");
+    }
+    let context = this;  
+    let currO = Object(this);
+  
+    if (arguments.length > 1) {
+      context = thisArg;
+    }
+  
+    if (typeof callback !== "function") {
+      throw new Error("Callback is not a function");
+    }
+  
+    let len = currO.length;  
+    let newArray = [];
+  
+    let i = 0;
+  
+    while (i < len) {
+      if (i in currO) {
+        newArray[i] = callback.call(context, this[i], i, currO);
+      }  
+      i++;
+    }
+  
+    return newArray;
+};
+Array.prototype.filter2 = function (callback, thisArg) {
+    if (this === null || this === void 0) {
+      throw new Error("Can't iterate over undefined or null");
+    }
+  
+    let context = this;  
+    let currO = Object(this);
+  
+    if (arguments.length > 1) {
+      context = thisArg;
+    }
+  
+    if (typeof callback !== "function") {
+      throw new Error("Callback is not a function");
+    }
+  
+    let currOLen = currO.length;  
+    let result = [];
+  
+    for (let i = 0; i < currOLen; i++) {
+      if (i in currO) {
+        let current = this[i];
+        if (callback.call(context, current, i, currO)) {
+            result.push(current);
+        }
+      }
+    }  
+    return result;
+};
+Array.prototype.reduce2 = function (callback, initValue) {
+    if (this === null) {
+      throw new Error("Cant iterate over undefined or null");
+    }
+  
+    let result;
+    let i = 0;
+  
+    let currO = Object(this);
+    let len = currO.length;
+  
+    if (typeof callback !== "function") {
+      throw new Error("Callback is not a function");
+    }
+  
+    if (arguments.length >= 2) {
+      result = initValue;
+    } else {
+      if (len === 0) {
+        throw new Error("Reduce of empty array with no initial value");
+      }
+      result = currO[i];
+      i++;
+    }
+  
+    for (; i < this.length; i++) {
+      if (i in currO) {
+        result = callback(result, currO[i], i, currO);
+      }
+    }
+  
+    return result;
+};
+
+const notes = [
```

Comment:
All the constants with the _collection_ meaning should be described with uppercase: `const NOTES = [`